doc/design.md

```diff
@@ -0,0 +1,949 @@
+However, for the interface of `API::MatrixWorkspace` (and in connection to that `HistogramData::Histogram`) this has led to an interface that is too complicated, with different API functions for read-only access, write-access, and access for sharing.
+
+
+### <a name="design-details-uncertainties-are-stored-as-variances"></a>Uncertainties are stored as variances, not standard deviations
+
+##### Rationale
+
+Mantid currently stores uncertainties as standard deviation (except for `DataObjects::MDHistoWorkspace`).
+However, most computations require the variance for propagation of uncertainties.
+To this end it is often squared in place (breaking assumptions about, e.g., `HistogramData::Histogram` and `DataObjects::Workspace2D`, and subsequently converted back to standard deviation by taking the square root.
+This also adds a little computational overhead.
+Having a clear and well defined strategy will reduce the risk of confusion.
+
+##### Implementation
+
+The go-to tag for storing uncertainties will be `Data::Variance`.
+`Data::StdDev` will exist, but is for *access* only, i.e., attempting to create a `Variable` with that tag should fail.
+Instead `Data::StdDev` can by used in combination with `DatasetView` to *read* the standard deviations if required, e.g., for visualization purposes.
+If it turns out to be useful it would also be possible to provide setters --- as opposed to a simple assignment which is possible if the actual underlying data is referenced --- for standard deviation using `DatasetView` .
+
+
+### <a name="design-details-distribution-histogram-distinguised-by-unit"></a>"Distributions" and "histogram data" are distinguished by their unit
+
+##### Rationale
+
+Currently `API::MatrixWorkspace` (and by extension `HistogramData::Histogram`) keep a flag indicating whether data in a histogram has been divided by its bin width.
+A considerable number of algorithms are checking this manually, branching to different code sections.
+`HistogramData::Histogram` attempted to solve this by introducing getters like `Histogram::counts()` and `Histogram::frequencies()`.
+However, it seems to be useful only in a rather limited number of places, whereas it adds a significant amount of complication to the API.
+The flag adds complication and should thus be removed.
+
+##### Implementation
+
+This will require having a unit for "counts", i.e., we cannot simply use "dimensionless".
+We need to establish whether it is possible to write clean client code based on the unit-driven mechanism.
+Apart from that there should not be any major implementation difficulties.
+
+
+### <a name="design-details-no-progress-reporting-no-cancellation"></a>No progress reporting and no cancellation
+
+##### Rationale
+
+- Most operations are simple and thus fast.
+- Cancellation and progress reporting can be implemented in higher-level algorithms that *use* `Dataset`.
+- Cancellation leads to undefined side effects for in-place operation, i.e., it breaks data.
```

Comment:
By extension then, all in-place operations will be `no-except`? I don't think it should be possible to recover/continue with a corrupted data structure.

Reply:
In fact, we should design an error policy for this library

Reply:
Probably it could be no-except, but since this is at a very high level I assume it would make very little difference?

Reply:
Regarding "error policy", is this covered by what I write below regarding exception guarantees? OR did you have anything else in mind?

Reply:
It's actually a little tricky in c++ at the moment to avoid using exceptions particularly in constructors, but I was thinking along the lines of errors in return values or any other considerations for how errors might be propagated.

Reply:
If it is an in-place operation there should be no (relevant) constructor calls.

Regarding error propagation, are you referring to something like storing NaN if a value could not be computed?